### ISSUE TITLE
[M-Perf-E4] Pipelined prefill for llama-engine (extend PR #265)

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/hydra_vortex#269]
## Summary

Add **Pipelined Prefill** to the llama-engine E1 architecture (PR #265). The current engine mode does synchronous P/D split — each request blocks through every stage (prefill → save → pick → restore → decode) before the next request can start. Pipelined prefill **overlaps stages across requests** so the decode GPU stays continuously busy.

**Target win:** 2-3x aggregate throughput at concurrency depth 2, 3-4x at depth 3 (Qwen35MoE: 4-12s prefill + 4s decode).

**Reference paper:** [arXiv:2504.08791](https://arxiv.org/abs/2504.08791) — Prima.cpp's Pipelined-Ring Parallelism (PRP) provides theoretical grounding for the async prefill pipeline pattern.

---

## Current State (PR #265, after E1)

```
Req A: [prefill]──[save]──[pick]──[restore]──[decode]
Req B:                                        [prefill]──[save]──[pick]──[restore]──[decode]
                                            ↑
                                    B starts only after A's decode slot is free
```

**Throughput bottleneck:** decode GPU idles 6-10s waiting for the next prefill to finish under sustained load. At 2-request concurrency, only ~50% of the decode GPU's potential is utilized.

**Engine mode uses 3 channels, 3 consumer tasks:**
- `_mainQueue` (size 500) — ingestion
- `_prefillQueue` (size 50) — PREFILL pipeline (one consumer per prefill worker)
- `_decodeQueue` (size 100) — DECODE pipeline (one consumer per decode worker)

Single-stage consumers block until terminal or handoff. **No overlap between prefill of N+1 and decode of N.**

---

## Target State (Pipelined)

```
Req A: [prefill]──[save]──[pick]──[restore]──[decode]
Req B:              [prefill]──[save]──[pick]──[restore]──[decode]
Req C:                          [prefill]──[save]──[pick]──[restore]──
                ↑                ↑                ↑
         B starts prefill   C starts prefill   A's decode slot
         while A saves        while B saves      still busy
```

**Throughput win** at scale:
- Concurrency 2: ~2x aggregate throughput
- Concurrency 4: ~3-3.5x aggregate throughput (Qwen35MoE 60K context)
- Warm follow-ups: skip PREFILL_BEGIN entirely (use PREFILL_RESUME with 0 chunks → microsecond latency)

---

### 1. New E1 RPC Opcodes (C++ + C#)

| Opcode | Hex | Purpose |
|--------|-----|---------|
| `HYDRA_OP_PREFILL_BEGIN` | `0x46` | Start prefill, return early after `n_tokens` to allow pipeline yield |
| `HYDRA_OP_PREFILL_CHUNK` | `0x47` | Stream partial KV + partial logits chunk (called by PREFILL_RESUME) |
| `HYDRA_OP_PREFILL_RESUME` | `0x48` | Continue prefill on same slot from token N, return next chunk |
| `HYDRA_OP_PREFILL_ABORT` | `0x49` | Cancel a started prefill, free the slot |

**C++ side** (`src/llama-cpp/tools/server/server-context.cpp`):
- `hydra_handle_prefill_begin` — tokenize prompt, prefill first chunk, return early with v3 mini-blob
- `hydra_handle_prefill_chunk` — return current chunk's logits + partial KV for the next n_ubatch tokens
- `hydra_handle_prefill_resume` — continue prefill from `last_n_past`, return next chunk
- `hydra_handle_prefill_abort` — clear slot, free lease

**C# side** (`src/core/Hydra.Shared/RpcClient.cs`):
```csharp
Task<RpcResponse> EnginePrefillBeginAsync(string slotId, string messagesJson, int firstChunkTokens, string traceId, CancellationToken ct);
Task<RpcResponse> EnginePrefillChunkAsync(string slotId, int fromNPast, int chunkTokens, string traceId, CancellationToken ct);
Task<RpcResponse> EnginePrefillResumeAsync(string slotId, int fromNPast, int chunkTokens, string traceId, CancellationToken ct);
Task<RpcResponse> EnginePrefillAbortAsync(string slotId, string traceId, CancellationToken ct);
```

### 2. New WorkItem States (`src/core/Hydra.Core/Models/WorkItem.cs`)

```
                  ┌── PrefillBegin
                  │      │
                  ▼      │
                PrefillChunk ◄──┐ (loop until n_prompt reached)
                  │             │
                  ▼             │
              PrefillMerge ────┘
                  │
                  ▼
            PrefillReadyForDecode   (KV in memory, waiting for decode slot)
                  │
                  ▼ (decode slot acquired)
              PickDecode
                  │
                  ▼
              RestoreKv
                  │
                  ▼
                Decode
```

**Critical new state:** `PrefillReadyForDecode` — KV assembled in memory, request waits in decode queue instead of prefill queue. This is the **pipeline buffer** that decouples stages.

State transitions encoded in `WorkItemState` enum:
```csharp
PrefillBegin = 28,        // start prefill on a worker
PrefillChunk = 29,        // fetch next chunk
PrefillMerge = 30,        // assemble final KV from chunks
PrefillReady = 31,        // ready for decode slot acquisition
DecodeQueued = 32,        // waiting in decode-acquire queue
```

### 3. Stage-Separated Consumers (`src/core/Hydra.Core/Services/WorkerSchedulerService.cs`)

**Current** (synchronous, 3 channels, 3 consumer types):
```csharp
_mainQueue      = Channel.CreateBounded<WorkItem>(500);  // ingestion
_prefillQueue   = Channel.CreateBounded<WorkItem>(50);   // prefill pipeline (1 per worker)
_decodeQueue    = Channel.CreateBounded<WorkItem>(100);  // decode pipeline (1 per worker)
```

**Target** (pipelined, 6 channels, 6 consumer types):
```csharp
_prefillBeginQueue   = Channel.CreateBounded<WorkItem>(50);   // start prefill on a worker
_prefillResumeQueue  = Channel.CreateBounded<WorkItem>(50);   // continue chunk
_prefillMergeQueue   = Channel.CreateBounded<WorkItem>(50);   // assemble KV from chunks
_decodeAcquireQueue  = Channel.CreateBounded<WorkItem>(200);  // wait for decode slot
_decodeActiveQueue   = Channel.CreateBounded<WorkItem>(100);  // active decode workers
```

**New `RunAsync` setup** in `WorkerSchedulerService.RunAsync`:
```csharp
// Stage 1: PREFILL_BEGIN — one consumer per prefill worker
RunPrefillBeginConsumer(_prefillBeginQueue.Reader, prefillSlots);
// Stage 2: PREFILL_RESUME — multiple consumers (2x per prefill worker)
RunPrefillResumeConsumer(_prefillResumeQueue.Reader, prefillSlots * 2);
// Stage 3: PREFILL_MERGE — single consumer
RunPrefillMergeConsumer(_prefillMergeQueue.Reader, 1);
// Stage 4: DECODE_ACQUIRE — one consumer per decode worker
RunDecodeAcquireConsumer(_decodeAcquireQueue.Reader, decodeSlots);
// Stage 5: DECODE_ACTIVE — one consumer per decode worker
RunDecodeActiveConsumer(_decodeActiveQueue.Reader, decodeSlots);
```

`IsPrefillPipeline()` partition becomes:
- **Begin stage:** `PrefillBegin, PrefillChunk` (returns when first chunk shipped)
- **Resume stage:** `PrefillChunk` (continues until full)
- **Merge stage:** `PrefillMerge, PrefillReady` (transitions to decode-acquire)
- **Decode stages:** `PickDecode, RestoreKv, Decode, BgSave, MarkEvicted, Done, ...`

### 4. Chunked KV Assembly (`WorkItem.cs` + `WorkerSchedulerService.cs`)

New field on `WorkItem`:
```csharp
public List<(int n_past, byte[] chunk_logits)> PrefillChunks { get; set; } = new();
public byte[]? MergedKvBlob { get; set; }  // assembled final blob, freed after Decode success
```

`PrefillMerge` state: iterate through `PrefillChunks`, concatenate logits (for activation handoff), build final v2 blob. This is **in-memory only** — Store is only hit if we need to migrate to a different decode worker.

**Chunk format (v3 mini-blob):**
```
[1B version=0x03][4B n_past][4B n_tok][n_tok*4B tokens][1B flags][?ckpt?][chunk_logits]
```

The `chunk_logits` is the partial logits tensor for the last token of the chunk — used for activation handoff. For warm follow-ups, `n_tok=0` and `chunk_logits` is empty.

### 5. Warm Session Fast Path

For warm sessions, the existing path (`KV restored via StatePut before decode`) is **already pipelined-compatible**:
- No PREFILL_BEGIN needed (KV already in memory)
- Skip straight to `EngineDecode` (existing path)
- Net win: zero latency cost for warm follow-ups in pipelined mode

Detection logic in `WorkerSchedulerService.RouteAsync`:
```csharp
if (entry?.NPast > 0 && entry.HasStoreState && !needsRefresh)
{
    // Warm fast path: skip PREFILL_BEGIN, go straight to PickDecode
    item.State = WorkItemState.PickDecode;
    return await PickDecodeAsync(item);
}
```

### 6. Slot Lease Integration

**Critical:** PREFILL_BEGIN acquires a slot lease. The lease must stay alive until `PrefillReady` → `Decode` completes (not until `PrefillBegin` returns).

```csharp
// In PrefillBegin: acquire lease, keep alive across stages
var lease = new SlotLease(worker.Name, slot, item.SessionId,
    LeaseLifetime.Short, _tracker);
// Lease.DisposeAsync() called only in FinalizeAsync after Done/Failed
```

For pipelined requests, **all in-flight requests on the same prefill worker** hold slot leases simultaneously. With 2 prefill workers and pipeline depth 2, you can have 2+2=4 simultaneous slot leases on the prefill GPU's pool.

### 7. Backpressure

Each channel has a bounded capacity. When a channel is full, `WriteAsync` awaits. This naturally applies backpressure upstream:
- If `_decodeActiveQueue` is full (decode workers saturated), `_decodeAcquireQueue.WriteAsync` blocks
- If `_decodeAcquireQueue` is full, `_prefillMergeQueue.WriteAsync` blocks
- If `_prefillMergeQueue` is full, `_prefillResumeQueue.WriteAsync` blocks
- If `_prefillResumeQueue` is full, `_prefillBeginQueue.WriteAsync` blocks
- If `_prefillBeginQueue` is full, `RunClassifierAsync` blocks

**Tunable per stage:** adjust channel sizes for workload profile. For chat: smaller decode queue. For long-doc: smaller prefill queue.

---

## Tests (`src/core/Tests.Core/Integration/EngineModeTests.cs`)

Add to the existing 8 tests:

| # | Test | Verifies |
|---|------|----------|
| 1 | `Pipelined_TwoRequests_PrefillOverlapsDecode` | Submit 2 requests; assert prefill of N+1 starts before decode of N |
| 2 | `Pipelined_AbortMidStream_FreesSlot` | Abort a prefill mid-chunk; assert slot returns to pool |
| 3 | `Pipelined_WarmSession_FastPath` | Warm follow-up uses `PREFILL_RESUME` (0 tokens) instead of full `PREFILL_BEGIN` |
| 4 | `Pipelined_StreamingDecode_OverRPC` | Exercise `EngineDecodeStreamAsync` (currently unused) end-to-end |
| 5 | `Pipelined_Backpressure_QueueFull` | Submit N+1 requests; assert backpressure propagates through channels |
| 6 | `Pipelined_ChunkOrdering_MonotonicNPast` | Verify chunks arrive in order (n_past strictly increasing) |
| 7 | `Pipelined_FailureRecovery_AbortedPicked` | Abort in-flight prefill, verify next request acquires slot |
| 8 | `Pipelined_StoreOnlyOnMigration` | Same-worker pipelined requests don't touch Store (KV stays in memory) |

**Update `tests/e1_rpc_test.py`:** send JSON `messages` (not raw token bytes) for PREFILL test, add chunked prefill tests.

---

## Files to Modify

| File | Change | LOC |
|------|--------|-----|
| `src/llama-cpp/tools/server/server-rpc.h` | Add 4 new opcodes (0x46-0x49) | 20 |
| `src/llama-cpp/tools/server/server-context.cpp` | Add 4 handler functions; refactor `hydra_handle_prefill` to support chunked streaming | 350 |
| `src/core/Hydra.Shared/Protocol.cs` | Add 4 new `OpCode` entries | 8 |
| `src/core/Hydra.Shared/RpcClient.cs` | Add 4 new RPC methods | 80 |
| `src/core/Hydra.Core/Models/WorkItem.cs` | Add 5 new states + 2 new fields (`PrefillChunks`, `MergedKvBlob`) | 30 |
| `src/core/Hydra.Core/Services/WorkerSchedulerService.cs` | Refactor 3 consumer types → 6; add stage-separated pipeline; warm fast-path | 350 |
| `src/core/Tests.Core/Integration/EngineModeTests.cs` | Add 8 new pipelined tests | 350 |
| `tests/e1_rpc_test.py` | Update to JSON payload; add chunked prefill tests | 100 |
| `specs/rpc-protocol.md` | Document 4 new E1 opcodes + v3 mini-blob wire format | 100 |
| `docs/milestone-perf.md` | Add E4 (Pipelined Prefill) section with diagrams | 80 |
| **Total** | | **~1470** |

---

## Estimated Effort

| Phase | Component | LOC | Days |
|-------|-----------|-----|------|
| **E3 (Foundation)** | Stage-separated consumers + new states + chunked prefill | 800 | 3.0 |
| **E4 (Streaming)** | `EngineDecodeStreamAsync` over RPC + true pipeline overlap tests | 670 | 4.0 |
| **Total** | | **~1470** | **~7 days** |

**Recommended split:**
- **Phase 1 (E3, ~3 days):** stage-separated consumers + chunked prefill. Tests use **non-pipelined** path (just refactor). Unblocks future work without taking on streaming risk.
- **Phase 2 (E4, ~4 days):** Add `EngineDecodeStreamAsync` over RPC + true pipeline overlap tests. Most of the throughput win comes from streaming decode.

---

## Risks

| Risk | Severity | Mitigation |
|------|----------|------------|
| Out-of-order chunk arrival if network reorders | High | Use slot-id in chunk RPC key, server validates monotonic n_past |
| Decode GPU KV cache pressure with pipelined requests | Medium | Bounded queue depth + per-session budget |
| Single-threaded C++ inference bottleneck (server_task queue) | High | Design with E2 in mind (per-slot task queues planned for future) |
| Slot reuse race: prefill slot reused before KV shipped to decode | High | Slot lease Long lifetime; DisposeAsync() only after Decode success |
| Warm session full re-prefill beats pipelined | Medium | Detect warm sessions, skip PREFILL_BEGIN entirely (use existing path) |
| In-flight prefill needs more than 2 GPU workers to demonstrate win | Low | Benchmark on single worker first; pipeline shows clear win at depth 2 |

---

## Related

- **Builds on:** PR #265 (HTTP server for llama-engine + E0/E1 scaffold)
- **Related to:** #161 (P/D split architecture), #266 (logits in state blob), #197 (C# Coordinator gaps)
- **Pre-req:** Two-Layer Slot Model (`feat/slot-lease-two-layer` branch) — slot lease Long/Short lifetimes
- **Reference:** arXiv:2504.08791 (Prima.cpp PRP), arXiv:2605.05219 (sparse prefix caching for hybrid)

---

## Handoff Notes for Implementers

1. **Start with the C++ side** — implement `hydra_handle_prefill_begin/chunk/resume/abort` first. Test via `tests/e1_rpc_test.py` before touching the C# scheduler.

2. **C# refactor before adding tests** — split the single `_prefillQueue` consumer into `_prefillBegin/Resume/Merge` consumers. This is a pure refactor; existing tests should still pass.

3. **Test chunked prefill in isolation** — write a C# test that uses `EngineTestRpcClient` to return canned chunks and verify the `PrefillChunks` list assembles correctly. Don't mix with real network yet.

4. **Add warm fast-path last** — once chunked prefill works, add the warm session detection that skips PREFILL_BEGIN entirely. This is the latency win for chat workloads.

5. **Watch the slot lease** — the lease for PREFILL_BEGIN must stay alive across all stages. The current `FinalizeAsync` calls `DisposeAsync()` on short-lived leases; pipelined requests may need explicit `DisposeAsync()` after `Decode` completes, not after `PrefillBegin`.

6. **Wire format details** — the v3 mini-blob differs from v2 in that it does NOT include the full KV state. Only the chunk's logits tensor. The final v2 blob (with full KV) is assembled in `PrefillMerge` from the partial chunks.
